frame/support/procedural/src/benchmark.rs

```diff
@@ -0,0 +1,836 @@
+
+						// consume #[extrinsic_call] tokens
+						expr_call.attrs.remove(k);
+
+						// extract origin from expr_call
+						let Some(origin) = expr_call.args.first().cloned() else {
+							return Some(Err(Error::new(expr_call.args.span(), "Single-item extrinsic calls must specify their origin as the first argument.")))
+						};
+
+						Some(Ok((i, BenchmarkCallDef::ExtrinsicCall { origin, expr_call, attr_span: attr.span() })))
+					})
+				},
+				Stmt::Expr(Expr::Block(block)) => { // #[block] case
+					block.attrs.iter().enumerate().find_map(|(k, attr)| {
+						let segment = attr.path.segments.last()?;
+						let _: keywords::block = syn::parse(segment.ident.to_token_stream().into()).ok()?;
+						let mut block = block.clone();
+
+						// consume #[block] tokens
+						block.attrs.remove(k);
+
+						Some(Ok((i, BenchmarkCallDef::Block { block, attr_span: attr.span() })))
+					})
+				},
+				_ => None
+			}
+		}).collect::<Result<Vec<_>>>()?;
+		let (i, call_def) = match &call_defs[..] {
+			[(i, call_def)] => (*i, call_def.clone()), // = 1
+			[] => return Err(Error::new( // = 0
+				item_fn.block.brace_token.span,
+				"No valid #[extrinsic_call] or #[block] annotation could be found in benchmark function body."
+			)),
+			_ => return Err(Error::new( // > 1
+				call_defs[1].1.attr_span(),
+				"Only one #[extrinsic_call] or #[block] attribute is allowed per benchmark."
+			)),
+		};
+
+		Ok(BenchmarkDef {
+			params,
+			setup_stmts: Vec::from(&item_fn.block.stmts[0..i]),
+			call_def,
+			verify_stmts: Vec::from(&item_fn.block.stmts[(i + 1)..item_fn.block.stmts.len()]),
+			extra,
+			skip_meta,
+		})
+	}
+}
+
+/// Parses and expands a `#[benchmarks]` or `#[instance_benchmarks]` invocation
+pub fn benchmarks(attrs: TokenStream, tokens: TokenStream, instance: bool) -> TokenStream {
+	let module = parse_macro_input!(tokens as ItemMod);
+	let mod_span = module.span();
+	let where_clause = match syn::parse::<Nothing>(attrs.clone()) {
+		Ok(_) => quote!(),
+		Err(_) => parse_macro_input!(attrs as WhereClause).predicates.to_token_stream(),
+	};
+	let mod_vis = module.vis;
+	let mod_name = module.ident;
+	let mut expanded_stmts: Vec<TokenStream2> = Vec::new();
+	let mut benchmark_defs: Vec<BenchmarkDef> = Vec::new();
+	let mut benchmark_names: Vec<Ident> = Vec::new();
+	let mut extra_benchmark_names: Vec<Ident> = Vec::new();
+	let mut skip_meta_benchmark_names: Vec<Ident> = Vec::new();
+	let Some((_brace, mut content)) = module.content else {
+		// this will compile error already because attributes attached to empty modules are unstable
+		// but including error anyway to make this future-proof
+		return quote_spanned!(mod_span=> "Module cannot be empty!".to_compile_error()).into()
+	};
+	for stmt in &mut content {
+		let mut push_stmt = || {
+			expanded_stmts.push(stmt.to_token_stream());
+		};
+		let Item::Fn(mut func) = stmt.clone() else { push_stmt(); continue; };
+		for (i, attr) in (&func.attrs.clone()).iter().enumerate() {
```

Comment:
Why does this need (&func.attrs.clone())? Looks to me that it can just be `func.attrs.iter().enumerate()`.

Reply:
Moreover, I feel like this loop can be changed into a `filter_map` and `fold`ed into a `(Vec<TokenStream2>, Vec<BenchmarkDef>, Vec<Ident>. Vec<Ident>, Vec<Ident>)`.

Reply:
No wait sorry, due to the `break` at the end of this loop, this is actually a `find_map` to a `BenchmarkDef`.

Reply:
> Why does this need (&func.attrs.clone())? Looks to me that it can just be `func.attrs.iter().enumerate()`.

Yup you are right, changed this one already, seeing if I can also do the `find_map`